src/DotNetWorker/Grpc/GrpcHttpRequestData.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Security.Claims;
+using Microsoft.Azure.WebJobs.Script.Grpc.Messages;
+
+namespace Microsoft.Azure.Functions.Worker
+{
+    internal class GrpcHttpRequestData : HttpRequestData
+    {
+        private readonly RpcHttp _httpData;
+        private Uri? _url;
+        private IEnumerable<ClaimsIdentity>? _identities;
+        private IImmutableDictionary<string, string> _headers;
+
+        public GrpcHttpRequestData(RpcHttp httpData)
+        {
+            _httpData = httpData ?? throw new ArgumentNullException(nameof(httpData));
+        }
+
+        public override ReadOnlyMemory<byte>? Body
+        {
+            get
+            {
+                // Based on the advertised worker capabilities, the payload should always be binary data
+                if (_httpData.Body.DataCase != TypedData.DataOneofCase.Bytes)
+                {
+                    throw new NotSupportedException($"{nameof(GrpcHttpRequestData)} expects binary data only. The provided data type was '{_httpData.Body.DataCase}'.");
+                }
+
+                return _httpData.Body.Bytes.Memory;
+            }
+        }
+
+        public override IImmutableDictionary<string, string> Headers => _headers ??= _httpData.NullableHeaders.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.Value);
```

Comment:
Do we need to make this something like `IHeaderDictionary`? I think @mhoeger mentioned that multiple identical keys wasn't supported... but maybe it could future-proof it?

Reply:
Done.